### PR TITLE
Neighbors dev

### PR DESCRIPTION
**Pull request by katesanders9**
Nearest neighbors functionality + featurization compatibility

### CODE REVIEW

File: cfg/tools/train_siamese.yaml [hunk 1]
```diff
@@ -3,30 +3,38 @@
 #####################################
```
Comment (visatish) [inline cfg/tools/train_siamese.yaml:3]: A few points about configs:
1. We usually try to keep a single example config, such as `train_siamese.yaml` in this case, that we don't modify. If you have a new config for nearest neighbors that is working, create a new one such as `train_siamese_nn.yaml`.
2. If this config isn't 100% working (like if you're just testing stuff), don't commit the changes.

This just helps us keep the configs in the master branch clean so that we can easily reference and replicate things.

File: memory/model/nearest_neighbors/neighbor_gen.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+from nearpy import Engine
```
Comment (visatish) [inline memory/model/nearest_neighbors/neighbor_gen.py:3]: For imports, I tend to follow the format of:
1) Libraries included in the standard Python installation such as `os`, `argparse`, etc.
3) [Space]
2) Third-party libraries such as `numpy` ,`tensorflow`, etc.
3) [Space]
4) External AUTOLab libraries such as `autolab_core`
5) [Space]
6) Internal imports from inside this library.

You can see an example [here](https://github.com/BerkeleyAutomation/memory/blob/master/memory/model/siamese/network.py#L8). This is just to keep things organized. If there's only 1 import in a few categories, feel free to merge things together.

File: memory/model/nearest_neighbors/neighbor_gen.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+from nearpy.hashes import RandomBinaryProjections
+from nearpy.filters import NearestFilter
+import numpy as np
+import os
+
+class NeighborGenerator:
+    def __init__(self, dimension=9984, num_neighbors=10, distance_fxn=None):
+        self.dimension = dimension
```
Comment (visatish) [inline memory/model/nearest_neighbors/neighbor_gen.py]: Try to use `_<variable_name>` for private variables and methods.

File: memory/model/nearest_neighbors/neighbor_gen.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+                image = os.path.join(obj_dir, file)
+                image = np.load(image)['arr_0']
+                self.engine.store_vector(image, obj)
+
+    def predict(self, file_location):
+        # takes in npz file from featurization script
+        image = np.load(file_location)['arr_0']
+        return self.engine.neighbours(image)
+
+    def batch_predict(self, dir):
```
Comment (visatish) [inline memory/model/nearest_neighbors/neighbor_gen.py]: Can this use `predict()`?

File: memory/model/siamese/network.py [hunk 5]
```diff
@@ -124,6 +124,8 @@ def _build_input_stream(self, input_node, layer_dict, name=None, layers=None):
             for layer_index, (layer_name, layer_config) in enumerate(layer_dict.items()):
                 layer_type = layer_config["type"]
 
+                print(layer_config.keys())
```
Comment (visatish) [inline memory/model/siamese/network.py]: Can you remove this?

File: memory/training/siamese/trainer.py [hunk 6]
```diff
@@ -17,8 +15,9 @@
 
 from memory.training.utils import (FileTemplates, DirTemplates,
                                    ImageDataset, DataGenerator,
-                                   build_contrastive_loss, GeneralConstants)
-
+                                   build_contrastive_loss)
```
Comment (visatish) [inline memory/training/siamese/trainer.py:20]: Can you add this import to `memory/model/nearest_neighbors/__init__.py` and then import it from there? I generally try to to import directly from other files unless they are in the same directory level.

File: memory/training/siamese/trainer.py [hunk 7]
```diff
@@ -175,3 +168,26 @@ def _setup(self):
         # build train and val data generators
         self._train_gen, self._val_gen = self._build_data_generators(train_dataset, val_dataset)
 
+
+    def neighbor_prediction(self):
```
Comment (visatish) [inline memory/training/siamese/trainer.py]: Can we move this to `SiameseNet`?

File: memory/training/utils/dataset.py [hunk 8]
```diff
@@ -9,6 +9,7 @@
 import random
 
 import skimage
+import skimage.io
```
Comment (visatish) [inline memory/training/utils/dataset.py:12]: This doesn't seem to be used.

File: memory/training/utils/dataset.py [hunk 9]
```diff
@@ -113,7 +114,7 @@ def generate_triple(self, pos=False):
             im2_views = os.path.join(folder1, view_str)
             #im2_views = os.path.join(folder1, "{}.npz".format(view_str))
             #im2_views = os.path.join(folder1, "view_000000")
-            im2 = os.path.join(im2_views, random.choice(os.listdir(im2_views)))
+            im2 = os.path.join(im2_views, random.choice(os.listdir(im2_views))  )
```
Comment (visatish) [inline memory/training/utils/dataset.py]: Can we remove the extra spaces in this and the following similar line?

File: tools/train_siamese.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
 from autolab_core import YamlConfig, Logger
 from memory import SiameseNet, SiameseTrainer
 
-# setup logger
+#setup logger
```
Comment (visatish) [inline tools/train_siamese.py]: This is super minor, but can we add the space back lol.

File: tools/train_siamese.py [hunk 11]
```diff
@@ -77,3 +77,6 @@
                              model_name=name)
     trainer.train()
     logger.info("Total Training Time:" + str(utils.get_elapsed_time(time.time() - start_time))) 
+
+    if train_config["neighbor_classification"]:
```
Comment (visatish) [inline tools/train_siamese.py]: I think it would be better to create another separate tool script such as `predict_siamese_nn.py`. It doesn't seem to be the best place to do it here because we don't seem to be actually using them, and also because this script should be solely for training. What I usually do is just copy an old script I have (such as `train_siamese.py` in this case), and then gut it out/modify it for the new thing I'm trying to do.

File: memory/training/siamese/trainer.py [hunk 12]
```diff
@@ -175,3 +168,26 @@ def _setup(self):
...
+        predictions = []
+
+        cache_dir = self.cache_dir
+        test_img_dir = self.test_dir
+
+        neighbor_gen = NeighborGenerator(self)
+        neighbor_gen.load_data(cache_dir)
+
+        # form [[img_file, [neighbors]], ...]
+        neighbors = neighbor_gen.batch_predict(test_img_dir)
```
Comment (visatish) [inline memory/training/siamese/trainer.py]: Can we make this into a more generic predict function that takes in a list of images and queries those images?

File: memory/training/siamese/trainer.py [hunk 13]
```diff
@@ -175,3 +168,26 @@ def _setup(self):
...
+
+        for pair in neighbors:
+            test_img = np.load(pair[0])['arr_0']
+            for neighbor in pair[1]:
+                neighbor = neighbor[1]
+                n_dir = os.path.join(cache_dir, neighbor)
+                neighbor_img = np.load(n_dir)['arr_0']
+
+                prediction = self._network.predict([np.array([test_img]), np.array([neighbor_img])])
+                predictions.append([pair[0], neighbor, prediction])
```
Comment (visatish) [inline memory/training/siamese/trainer.py]: It would be nice if dim 0 of the returned predictions matched up with dim 0 of the input queries. For example, if there are 10 input images, we would look up 10 neighbors for each one and get 10 predictions, so our output shape would be (10, 10). I don't think we need to return the actual query images (in this case coming from `test_img_dir`), because in the more generic predict function format I specified above, the user will anyways be providing those. Also, I would recommend using numpy arrays instead of lists, especially when you have larger arrays, because then you can take advantage of vectorized numpy operations. Usually what I do is allocate an array of zeros and then fill it in as I go. This is usually faster than things like `hstack` because you don't have to copy memory when you have to resize.

File: memory/model/nearest_neighbors/neighbor_gen.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        # expects dir to be in same format that siamese network uses, but no view folder
+        # takes in npz files produced by featurization script
+        for obj in os.listdir(dir):
+            obj_dir = os.path.join(os.path.join(dir, obj), 'view_000000')
+            for file in os.listdir(obj_dir):
+                image = os.path.join(obj_dir, file)
+                image = np.load(image)['arr_0']
+                self.engine.store_vector(image, obj)
+
+    def predict(self, file_location):
```
Comment (visatish) [inline memory/model/nearest_neighbors/neighbor_gen.py]: Can we have this and the following `batch_predict` take in arrays of inputs to query instead of directories? In general, try to keep operations on directories at the highest level of your code (such as in the `tools/` scripts), and use array abstractions for lower levels.

File: memory/training/utils/dataset.py (hunk 8 above)
Comment (katesanders9) [reply memory/training/utils/dataset.py:12]: At least on my machine, when `skimage` is imported it doesn't automatically import the `io` module---that has to be imported separately. I can't remember where exactly but I think `skimage.io` definitely is used somewhere in the code because python yelled at me when I tried to run the script without importing the extra module.

File: memory/training/siamese/trainer.py (hunk 6 above)
Comment (katesanders9) [reply memory/training/siamese/trainer.py:20]: Sorry, I'm a bit confused by this. Are you talking about the entire `from memory.training.utils import (FileTemplates, ... ,build_contrastive_loss)` import? I'm not sure why we would want to put that specifically in the `nearest_neighbors/__init__.py` script.
